Core/Filesystem/nand_core.c

```diff
@@ -0,0 +1,366 @@
+ */\
+#include <stdint.h>
+#include <string.h>
+#include "nandfs.h"
+#include "nand_core.h"
+#include "nand_m79a_lld.h"
+#include "nand_types.h"
+#include "nand_errno.h"
+
+static int lowest_inode = 0;
+static int highest_inode = 0;
+
+#define UINT32_MAX 0xFFFFFFFF;
+
+/*
+ * Finds the first unformatted block
+ * @Return:
+ * 0, success, blank space found
+ * -1, failed. Could be I/O or full device, No distinction made
+ */
+int _find_blank(PhysicalAddrs *addr) {
+	PhysicalAddrs search = {0};
+	inode_t node = {0};
+	NAND_ReturnType status = Ret_Failed;
+	for (uint16_t i = 0; i < NUM_BLOCKS; i++) {
+		search.block = i;
+		search.plane = i & 1;
+		status = NAND_Page_Read(&search, sizeof(node), (uint8_t *)&node);
+		if (status != Ret_Success) {
+			continue; // Might have hit a bad block, not really sure
+		}
+		if (node.magic != MAGIC) {
+			memcpy(addr, &search, sizeof(PhysicalAddrs));
+			return 0;
+		}
+	}
+	return -1;
+}
+
+/*
+ * Finds the PhysicalAddr of the first block of an inode by its ID
+ */
+static int _find_inode(PhysicalAddrs *addr, int inodeid) {
+	PhysicalAddrs search = {0};
+	inode_t node = {0};
+	NAND_ReturnType status = Ret_Failed;
+	for (uint16_t i = 0; i < NUM_BLOCKS; i++) {
+		search.block = i;
+		search.plane = i & 1;
+		status = NAND_Page_Read(&search, sizeof(node), (uint8_t *)&node);
+		if (status != Ret_Success) {
+			continue; // Might have hit a bad block, not really sure
+		}
+		if (node.id == inodeid && node.isfirst) {
+			memcpy(addr, &search, sizeof(PhysicalAddrs));
+			return 0;
+		}
+	}
+	return -1;
+}
+
+int _find_lowest_inode() {
+	uint32_t lowest = UINT32_MAX;
+	PhysicalAddrs search = {0};
+	inode_t node = {0};
+	NAND_ReturnType status = Ret_Failed;
+	int found_valid_file = 0; // Sanity check, if no files found, we will return 0;
+	for (uint16_t i = 0; i < NUM_BLOCKS; i++) {
+		search.block = i;
+		search.plane = i & 1;
+		status = NAND_Page_Read(&search, sizeof(node), (uint8_t *)&node);
+		if (status != Ret_Success) {
+			continue; // Might have hit a bad block, not really sure
+		}
+		if (node.magic != MAGIC) {
+			continue;
+		}
+		found_valid_file = 1;
+		if (node.id < lowest) {
+			lowest = node.id;
+		}
+	}
+	if (found_valid_file) {
+		return lowest;
+	}
+	return 0;
+}
+
+int _find_highest_inode() {
+	uint32_t highest = 0;
+	PhysicalAddrs search = {0};
+	inode_t node = {0};
+	NAND_ReturnType status = Ret_Failed;
+	int found_valid_file = 0; // Sanity check, if no files found, we will return 0;
+	for (uint16_t i = 0; i < NUM_BLOCKS; i++) {
+		search.block = i;
+		search.plane = i & 1;
+		status = NAND_Page_Read(&search, sizeof(node), (uint8_t *)&node);
+		if (status != Ret_Success) {
+			continue; // Might have hit a bad block, not really sure
+		}
+		if (node.magic != MAGIC) {
+			continue;
+		}
+		found_valid_file = 1;
+		if (node.id > highest) {
+			highest = node.id;
+		}
+	}
+	if (found_valid_file) {
+		return highest;
+	}
+	return 0;
+}
+
+int NANDfs_Core_Init() {
+    NAND_Init();
```

Comment:
i n d e n t a t i o n

Reply:
Rewrote history so this comment never existed